agent db: make rejecting ooo samples configurable

rabenhorst opened the pull request:
Add `storage.agent.reject-out-of-order-samples` flag and make rejecting out-of-order samples in agent's db configurable. This allows disabling remote writing out-of-order samples from agent, which was introduced in https://github.com/prometheus/prometheus/pull/12897.

ArthurSens commented:
Maybe I'm missing something, but isn't it already possible to disable OOO ingestion with the configuration file?

From my understanding if `out_of_order_time_window` is zero (the default), then OOO is disabled. See https://prometheus.io/docs/prometheus/latest/configuration/configuration/#tsdb

rabenhorst commented:
> From my understanding if `out_of_order_time_window` is zero (the default), then OOO is disabled. See https://prometheus.io/docs/prometheus/latest/configuration/configuration/#tsdb

I don't see how this applies to the agent db, which, as I understand, is just the TSDB WAL.

sdufel commented:
When agents ingest out-of-order samples and and remote write them to external stores, it can be extremely expensive on the other end.  We run a setup with prometheus agents writing to thanos receivers, and out of order samples generate severe amounts of load.

ArthurSens commented:
> Maybe I miss something, but I don't see how this applies to the agent db, which, as I understand, is just the TSDB WAL.

Yeah, I'm reading the code now and it looks like the agent indeed doesn't take this into consideration. 

I think that creating yet another configuration option for out of order would be confusing though. Could we re-use the already existing `out_of_order_time_window` for that?

rabenhorst commented:
> Yeah, I'm reading the code now and it looks like the agent indeed doesn't take this into consideration.
> 
> I think that creating yet another configuration option for out of order would be confusing though. What do you think about re-using the already existing `out_of_order_time_window` configuration option?

I think it makes sense and I changed it and added some test cases for edge cases. Is this what you had in mind? 

It might be confusing now that `out_of_order_time_window` is also applied to agent db.  What would be the best place to document it?

ArthurSens commented on cmd/prometheus/main.go:
```diff
@@ -1180,6 +1180,7 @@ func main() {
 	if agentMode {
 		// WAL storage.
 		opts := cfg.agent.ToAgentOptions()
+		opts.SetOutOfOrderTimeWindow(cfg.tsdb.OutOfOrderTimeWindow)
```
Hmmm, maybe we could remove this new line and set the out-of-order window inside `ToAgentOptions`?

Not sure if that's nitpicking or not to be honest, what do you think?

rabenhorst replied:
Done. Is that what you had in mind?

ArthurSens commented on tsdb/agent/db.go:
```diff
@@ -81,19 +81,30 @@ type Options struct {
+	OutOfOrderTimeWindow int64
+}
+
+func (o *Options) SetOutOfOrderTimeWindow(outOfOrderTimeWindow int64) {
+	if outOfOrderTimeWindow < 0 {
+		return
+	}
+	o.OutOfOrderTimeWindow = outOfOrderTimeWindow
 }
```
I see that you also use this function in tests, but we could simply use `opts.OutOfOrderTimeWindow = <somevalue>` over there

ArthurSens commented on tsdb/agent/db_test.go:
```diff
@@ -879,6 +882,61 @@ func TestDBAllowOOOSamples(t *testing.T) {
+			s := createTestAgentDB(t, reg, opts)
+			app := s.Appender(context.TODO())
+
+			var expectedErr error
+			if c.reject {
+				expectedErr = storage.ErrOutOfOrderSample
+			}
```
What do you think about changing `reject bool` to `error`, that could be either unset(nil) or `storage.ErrOutOfOrderSample` when declaring the test cases?

This way we could delete this if, just a little less code to read 😛

ArthurSens replied:
We wouldn't be able to delete the if below, but that's fine I guess

rabenhorst replied:
Makes sense. Done!

ArthurSens reviewed (commented):
It LGTM, just a few nitpicks

I'll try to ping maintainers to take a look at this as well

ArthurSens commented:
I'm also realizing that changing the default behavior now (always accept OOO -> have to configure window to accept) might be a breaking change. Since both agent-mode is a feature flag and OOO is marked as experimental in our docs, I think the change should be fine, but let's see what others think

ArthurSens commented:
> It might be confusing now that `out_of_order_time_window`, which is part of tsdb config, is also applied to agent db.  What would be the best place to document it?

Good point 😬, I don't remember if the configuration documentation is auto-generated... I'm 90% sure it's not. So it should be fine to just add a statement [here](https://github.com/prometheus/prometheus/blob/main/docs/configuration/configuration.md#:~:text=%23%20Configures%20how,of%2Dorder%20window.) clarifying that it is also applied for agent-mode?

rabenhorst commented:
> It LGTM, just a few nitpicks
> 
> I'll try to ping maintainers to take a look at this as well

I addressed the nits and thx!

roidelapluie commented:
cc @rfratto

rabenhorst commented:
> Since we're reusing an existing Prometheus config field, should we update docs for `out_of_order_time_window` around https://prometheus.io/docs/prometheus/latest/configuration/configuration/#tsdb to mention this dual use-case?

I added it to `out_of_order_time_window` docs: https://github.com/prometheus/prometheus/blob/017b43ec69b6113c3e4788a07f3936b597b386af/docs/configuration/configuration.md?plain=1#L3805-L3807

WDYT?

ArthurSens reviewed (approved):
Since the next release process starts next week, I'd merge this even without @rfratto's approval 😬.

@rabenhorst, it seems like you forgot to sign your lastest commit. Could you fix that please? 🙂

rabenhorst commented:
> Since the next release process starts next week, I'd merge this even without @rfratto's approval 😬.
> 
> @rabenhorst, it seems like you forgot to sign your lastest commit. Could you fix that please? 🙂

Done.

ArthurSens commented:
Thanks @rabenhorst

rfratto reviewed (commented):
Sorry for the post-merge review and holding this up longer than it should've been, but this looks good to me!
